File: sdk/go/webhooks.go
package qeetid

import (
	"context"
	"net/http"
	"net/url"
)

type Webhook struct {
	ID        string   `json:"id"`
	TenantID  string   `json:"tenant_id"`
	URL       string   `json:"url"`
	Events    []string `json:"events"`
	Enabled   bool     `json:"enabled"`
	Secret    string   `json:"secret,omitempty"`
	CreatedAt string   `json:"created_at"`
	UpdatedAt string   `json:"updated_at,omitempty"`
}

type CreateWebhookInput struct {
	URL     string   `json:"url"`
	Events  []string `json:"events"`
	Enabled *bool    `json:"enabled,omitempty"`
}

type UpdateWebhookInput struct {
	URL     *string  `json:"url,omitempty"`
	Events  []string `json:"events,omitempty"`
	Enabled *bool    `json:"enabled,omitempty"`
}

type WebhookDelivery struct {
	ID             string `json:"id"`
	WebhookID      string `json:"webhook_id"`
	Event          string `json:"event"`
	Status         string `json:"status"`
	ResponseStatus int    `json:"response_status,omitempty"`
	CreatedAt      string `json:"created_at"`
}

type Webhooks struct{ http *httpClient }

func (r *Webhooks) Create(ctx context.Context, tenantID string, in CreateWebhookInput) (*Webhook, error) {
	var out Webhook
	err := r.http.do(ctx, http.MethodPost, "/v1/tenants/"+url.PathEscape(tenantID)+"/webhooks", nil, in, &out, false)
	return &out, err
}

func (r *Webhooks) Get(ctx context.Context, tenantID, id string) (*Webhook, error) {
	var out Webhook
	err := r.http.do(ctx, http.MethodGet, "/v1/tenants/"+url.PathEscape(tenantID)+"/webhooks/"+url.PathEscape(id), nil, nil, &out, true)
	return &out, err
}

func (r *Webhooks) Update(ctx context.Context, tenantID, id string, in UpdateWebhookInput) (*Webhook, error) {
	var out Webhook
	err := r.http.do(ctx, http.MethodPatch, "/v1/tenants/"+url.PathEscape(tenantID)+"/webhooks/"+url.PathEscape(id), nil, in, &out, false)
	return &out, err
}

func (r *Webhooks) Delete(ctx context.Context, tenantID, id string) error {
	return r.http.do(ctx, http.MethodDelete, "/v1/tenants/"+url.PathEscape(tenantID)+"/webhooks/"+url.PathEscape(id), nil, nil, nil, true)
}

func (r *Webhooks) Test(ctx context.Context, tenantID, id string) error {
	return r.http.do(ctx, http.MethodPost, "/v1/tenants/"+url.PathEscape(tenantID)+"/webhooks/"+url.PathEscape(id)+"/test", nil, struct{}{}, nil, false)
}

func (r *Webhooks) List(ctx context.Context, tenantID string) ([]Webhook, error) {
	var env struct {
		Items []Webhook `json:"items"`
		Data  []Webhook `json:"data"`
	}
	if err := r.http.do(ctx, http.MethodGet, "/v1/tenants/"+url.PathEscape(tenantID)+"/webhooks", nil, nil, &env, true); err != nil {
		return nil, err
	}
	if env.Items != nil {
		return env.Items, nil
	}
	return env.Data, nil
}

func (r *Webhooks) ListDeliveries(ctx context.Context, tenantID, webhookID string) ([]WebhookDelivery, error) {
	var env struct {
		Items []WebhookDelivery `json:"items"`
		Data  []WebhookDelivery `json:"data"`
	}
	path := "/v1/tenants/" + url.PathEscape(tenantID) + "/webhooks/" + url.PathEscape(webhookID) + "/deliveries"
	if err := r.http.do(ctx, http.MethodGet, path, nil, nil, &env, true); err != nil {
		return nil, err
	}
	if env.Items != nil {
		return env.Items, nil
	}
	return env.Data, nil
}

func (r *Webhooks) RetryDelivery(ctx context.Context, tenantID, webhookID, deliveryID string) error {
	path := "/v1/tenants/" + url.PathEscape(tenantID) + "/webhooks/" + url.PathEscape(webhookID) + "/deliveries/" + url.PathEscape(deliveryID) + "/retry"
	return r.http.do(ctx, http.MethodPost, path, nil, struct{}{}, nil, false)
}

File: sdk/js/node/src/webhooks.ts
import type { HttpClient } from "./client.js";

export interface Webhook {
  id: string;
  tenant_id: string;
  url: string;
  events: string[];
  enabled: boolean;
  secret?: string;
  created_at: string;
  updated_at?: string;
}

export interface CreateWebhookInput {
  url: string;
  events: string[];
  enabled?: boolean;
}

export interface UpdateWebhookInput {
  url?: string;
  events?: string[];
  enabled?: boolean;
}

export interface WebhookDelivery {
  id: string;
  webhook_id: string;
  event: string;
  status: string;
  response_status?: number;
  created_at: string;
}

interface ListEnvelope<T> {
  items?: T[];
  data?: T[];
}

export class Webhooks {
  constructor(private readonly http: HttpClient) {}

  create(tenantId: string, input: CreateWebhookInput): Promise<Webhook> {
    return this.http.post<Webhook>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/webhooks`,
      input,
    );
  }

  get(tenantId: string, id: string): Promise<Webhook> {
    return this.http.get<Webhook>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/webhooks/${encodeURIComponent(id)}`,
    );
  }

  update(tenantId: string, id: string, input: UpdateWebhookInput): Promise<Webhook> {
    return this.http.patch<Webhook>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/webhooks/${encodeURIComponent(id)}`,
      input,
    );
  }

  delete(tenantId: string, id: string): Promise<void> {
    return this.http.delete<void>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/webhooks/${encodeURIComponent(id)}`,
    );
  }

  test(tenantId: string, id: string): Promise<void> {
    return this.http.post<void>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/webhooks/${encodeURIComponent(id)}/test`,
      {},
    );
  }

  async list(tenantId: string): Promise<Webhook[]> {
    const res = await this.http.get<ListEnvelope<Webhook>>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/webhooks`,
    );
    return res.items ?? res.data ?? [];
  }

  async listDeliveries(tenantId: string, webhookId: string): Promise<WebhookDelivery[]> {
    const res = await this.http.get<ListEnvelope<WebhookDelivery>>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/webhooks/${encodeURIComponent(webhookId)}/deliveries`,
    );
    return res.items ?? res.data ?? [];
  }

  retryDelivery(tenantId: string, webhookId: string, deliveryId: string): Promise<void> {
    return this.http.post<void>(
      `/v1/tenants/${encodeURIComponent(tenantId)}/webhooks/${encodeURIComponent(webhookId)}/deliveries/${encodeURIComponent(deliveryId)}/retry`,
      {},
    );
  }
}
